ortools/math_opt/solvers/xpress_solver.cc

```diff
@@ -0,0 +1,693 @@
+    const VariablesProto& new_variables) {
+  const int num_new_variables = new_variables.lower_bounds().size();
+  std::vector<char> variable_type(num_new_variables);
+  int n_variables = xpress_->GetNumberOfVariables();
+  for (int j = 0; j < num_new_variables; ++j) {
+    const VarId id = new_variables.ids(j);
+    InsertOrDie(&variables_map_, id, j + n_variables);
+    variable_type[j] =
+        new_variables.integers(j) ? XPRS_INTEGER : XPRS_CONTINUOUS;
+    if (new_variables.integers(j)) {
+      is_mip_ = true;
+      return absl::UnimplementedError("XpressSolver does not handle MIPs yet");
+    }
+  }
+  RETURN_IF_ERROR(xpress_->AddVars({}, new_variables.lower_bounds(),
+                                   new_variables.upper_bounds(),
+                                   variable_type));
+
+  // Not adding names for performance (have to call XPRSaddnames)
+  // TODO: keep names in a cache and add them when needed?
+
+  return absl::OkStatus();
+}
+
+XpressSolver::XpressSolver(std::unique_ptr<Xpress> g_xpress)
+    : xpress_(std::move(g_xpress)) {}
+
+absl::Status XpressSolver::AddNewLinearConstraints(
+    const LinearConstraintsProto& constraints) {
+  const int num_new_constraints = constraints.lower_bounds().size();
+
+  std::vector<char> constraint_sense;
+  std::vector<double> constraint_rhs;
+  std::vector<double> constraint_rng;
+  std::vector<LinearConstraintData*> new_slacks;
+  constraint_rhs.reserve(num_new_constraints);
+  constraint_sense.reserve(num_new_constraints);
+  new_slacks.reserve(num_new_constraints);
+  int n_constraints = xpress_->GetNumberOfConstraints();
+  for (int i = 0; i < num_new_constraints; ++i) {
+    const int64_t id = constraints.ids(i);
+    LinearConstraintData& constraint_data =
+        InsertKeyOrDie(&linear_constraints_map_, id);
+    const double lb = constraints.lower_bounds(i);
+    const double ub = constraints.upper_bounds(i);
+    constraint_data.lower_bound = lb;
+    constraint_data.upper_bound = ub;
+    constraint_data.constraint_index = i + n_constraints;
+    char sense = XPRS_EQUAL;
+    double rhs = 0.0;
+    double rng = 0.0;
+    const bool lb_is_xprs_neg_inf = lb <= XPRS_MINUSINFINITY;
+    const bool ub_is_xprs_pos_inf = ub >= XPRS_PLUSINFINITY;
+    if (lb_is_xprs_neg_inf && !ub_is_xprs_pos_inf) {
+      sense = XPRS_LESS_EQUAL;
+      rhs = ub;
+    } else if (!lb_is_xprs_neg_inf && ub_is_xprs_pos_inf) {
+      sense = XPRS_GREATER_EQUAL;
+      rhs = lb;
+    } else if (lb == ub) {
+      sense = XPRS_EQUAL;
+      rhs = lb;
+    } else {
+      sense = XPRS_RANGE;
+      rhs = ub;
+      rng = ub - lb;
+    }
+    constraint_sense.emplace_back(sense);
+    constraint_rhs.emplace_back(rhs);
+    constraint_rng.emplace_back(rng);
+  }
+  // Add all constraints in one call.
+  return xpress_->AddConstrs(constraint_sense, constraint_rhs, constraint_rng);
+}
+
+absl::Status XpressSolver::AddSingleObjective(const ObjectiveProto& objective) {
```

Comment:
What is this function supposed to do? At the moment (unless I am mistaken) it only changes the coefficients of variables that appear in `objective`. If a variable is not in `objective` then its objective coefficient won't be changed. In particular, it would not be zeroed. Or is `objective` always dense?